File: code-for-my-graduate-paper-cpp-version/src/utils/superpixel_utils.cpp
#include "../../include/superpixel_utils.h"
#include <opencv2/ximgproc.hpp>
#include <fstream>

cv::Mat SuperpixelUtils::generate_superpixels(const cv::Mat& image, int n_segments, double compactness) {
    cv::Mat labels;
    cv::Mat img_float;
    image.convertTo(img_float, CV_32FC3);
    
    auto slic = cv::ximgproc::createSuperpixelSLIC(img_float, cv::ximgproc::SLIC, n_segments, compactness);
    slic->iterate();
    slic->getLabels(labels);
    
    labels += 1; // Convert to 1-based indexing
    return labels;
}

cv::Mat SuperpixelUtils::read_dat_file(const cv::Size& image_size, const std::string& data_path) {
    std::ifstream file(data_path, std::ios::binary);
    std::vector<uint32_t> data(image_size.width * image_size.height);
    file.read(reinterpret_cast<char*>(data.data()), data.size() * sizeof(uint32_t));
    
    cv::Mat labels(image_size.height, image_size.width, CV_32S);
    for (int i = 0; i < image_size.height; i++) {
        for (int j = 0; j < image_size.width; j++) {
            labels.at<int>(i, j) = data[j * image_size.height + i] + 1;
        }
    }
    return labels;
}

void SuperpixelUtils::save_dat_file(const cv::Mat& labels, const std::string& data_path) {
    std::ofstream file(data_path, std::ios::binary);
    for (int j = 0; j < labels.cols; j++) {
        for (int i = 0; i < labels.rows; i++) {
            uint32_t val = labels.at<int>(i, j) - 1;
            file.write(reinterpret_cast<const char*>(&val), sizeof(uint32_t));
        }
    }
}

File: code-for-my-graduate-paper-python-version/superpixel_utils.py
"""
Superpixel utilities
"""
import numpy as np
from skimage.segmentation import slic
import struct

def generate_superpixels(image, n_segments=200, compactness=20):
    """
    Generate superpixels using SLIC algorithm
    Args:
        image: input image (H, W, 3)
        n_segments: number of superpixels
        compactness: compactness parameter
    Returns:
        superpixel labels (H, W)
    """
    # Convert to 0-1 range if needed
    if image.max() > 1:
        image = image / 255.0
    
    # Generate superpixels (labels start from 0)
    labels = slic(image, n_segments=n_segments, compactness=compactness, 
                  start_label=1, enforce_connectivity=True)
    
    return labels

def read_dat_file(image_size, data_path):
    """
    Read superpixel labels from .dat file
    Args:
        image_size: (height, width)
        data_path: path to .dat file
    Returns:
        label matrix (height, width)
    """
    height, width = image_size
    
    with open(data_path, 'rb') as f:
        # Read uint32 values
        data = f.read(height * width * 4)  # 4 bytes per uint32
        labels = struct.unpack(f'{height * width}I', data)
    
    # Convert to numpy array and add 1 (MATLAB indexing)
    labels = np.array(labels) + 1
    
    # Reshape to image dimensions
    labels = labels.reshape(width, height).T
    
    return labels

def save_dat_file(labels, data_path):
    """
    Save superpixel labels to .dat file
    Args:
        labels: label matrix (height, width)
        data_path: output path
    """
    # Convert to 0-based indexing and flatten
    labels_flat = (labels - 1).T.flatten().astype(np.uint32)
    
    with open(data_path, 'wb') as f:
        f.write(labels_flat.tobytes())